go/vt/mysqlctl/capabilities.go

```diff
@@ -0,0 +1,248 @@
+func (mysqld *Mysqld) IsMariaDB() bool {
+	return mysqld.flavor == flavorMariaDB
+}
+
+// MajorVersion returns an integer for the leading version.
+// For example 8 in the case of MySQL 8.0.11, or
+// 5 in the case of 5.6.15
+func (mysqld *Mysqld) MajorVersion() int {
+	return mysqld.version[0]
+}
+
+// MinorVersion returns an integer for the second version
+// For example 6 in the case of 5.6.11
+func (mysqld *Mysqld) MinorVersion() int {
+	return mysqld.version[1]
+}
+
+// PatchVersion returns an integer for the third
+// part of the version. For example 11 in the case of
+// 5.6.11
+func (mysqld *Mysqld) PatchVersion() int {
+	return mysqld.version[2]
+}
+
+func (mysqld *Mysqld) getVersionString() (string, error) {
+
+	mysqlRoot, err := vtenv.VtMysqlRoot()
+	if err != nil {
+		return "", err
+	}
+	mysqldPath, err := binaryPath(mysqlRoot, "mysqld")
+	if err != nil {
+		return "", err
+	}
+
+	_, version, err := execCmd(mysqldPath, []string{"--version"}, nil, mysqlRoot, nil)
+	if err != nil {
+		return "", err
+	}
+
+	return version, nil
+
+}
+
+func (mysqld *Mysqld) detectFlavor() (string, error) {
+
+	mysqlFlavor := os.Getenv("MYSQL_FLAVOR")
+	if len(mysqlFlavor) > 0 {
+		mysqlFlavor = strings.ToLower(mysqlFlavor)[0:5]
+
+		if mysqlFlavor == "mysql" {
+			return flavorMySQL, nil
+		} else if mysqlFlavor == "maria" {
+			return flavorMariaDB, nil
+		}
+	}
+
+	// Flavor was either not defined in ENV, or unknown
+
+	version, err := mysqld.getVersionString()
+	if err != nil {
+		return version, err
+	}
+
+	// Check if the flavor has been defined
+	// This takes first precedence.
+
+	if strings.Contains(version, "MySQL Community Server") {
+		return flavorMySQL, nil
+	} else if strings.Contains(version, "Percona") {
+		return flavorPercona, nil
+	} else if strings.Contains(version, "MariaDB") {
+		return flavorMariaDB, nil
+	}
+
+	return "", fmt.Errorf("Could not autodetect flavor!")
+
+}
+
+func (mysqld *Mysqld) detectVersion() (serverVersion, error) {
+
+	mysqlFlavor := strings.ToLower(os.Getenv("MYSQL_FLAVOR"))
+
+	// Try not to be too smart with version parsing.
+	// Assume GA versions
+	switch mysqlFlavor {
+	case "mysql56":
+		return serverVersion{5, 6, 10}, nil
+	case "mysql57":
+		return serverVersion{5, 7, 9}, nil
+	case "mysql80":
+		return serverVersion{8, 0, 11}, nil
+	case "mariadb10":
+		return serverVersion{10, 0, 10}, nil
+	case "mariadb101":
+		return serverVersion{10, 1, 8}, nil
+	case "mariadb102":
+		return serverVersion{10, 2, 6}, nil
+	case "mariadb103":
+		return serverVersion{10, 3, 7}, nil
+	case "mariadb104":
+		return serverVersion{10, 4, 6}, nil
+	}
+
+	// Could not understand version from ENV
+	// Time to auto-detect instead
+
+	version, err := mysqld.getVersionString()
+	if err != nil {
+		return nil, err
+	}
+
+	// Detect server version
+	re := regexp.MustCompile(`Ver [0-9]+\.[0-9]+\.[0-9]+`)
+	ver := re.Find([]byte(version))[4:]
+	re = regexp.MustCompile(`[0-9]+`)
```

Comment:
We can use submatches to capture the parts of the version in one go, instead of a second expression. Also we can use the String forms directly instead of the byte forms:

```go
re := regexp.MustCompile(`Ver ([0-9]+)\.([0-9]+)\.([0-9]+)`)
v := re.FindStringSubmatch(version)
if len(v) != 4 {
  return serverVersion{}, fmt.Errorf("unexpected mysqld --version format: %v", version)
}
major, _ := strconv.Atoi(string(v[1]))
minor, _ := strconv.Atoi(string(v[2]))
...
```

Notice we skip v[0] since that's the full match.